### PR TITLE
Send Customer.io notifications and warnings for connections being auto-disabled

### LINKED ISSUE #10929: Notify users of auto-disabled connections
## Tell us about the problem you're trying to solve
With the additional feature of adding an option to auto-disable failing connections from this [issue](https://github.com/airbytehq/airbyte/issues/9715), we want to clearly define how the notification of a connection being auto-disabled should be surfaced.

For example, do we want an email or a pop-up? Do we want to confirm that the user has acknowledged this automated disablement of a connection? etc.

### PR DESCRIPTION
**Pull request by terencecho**
## What
A [feature flag](https://github.com/airbytehq/airbyte/issues/9715) exists where connections that have been constantly failing will be auto-disabled to save on resources. This PR utilizes infrastructure from a [previous PR](https://github.com/airbytehq/airbyte/pull/11220) to send notifications of when connections have been auto-disabled, or warnings that they are about to be auto-disabled.

These notifications require that a `Customer.io` api key is set as an env variable and that the `AUTO_DISABLE_FAILING_CONNECTIONS` flag is turned on. It also makes the assumption that an email template exists in your Customer.io account.

This PR address issue [#10929](https://github.com/airbytehq/airbyte/issues/10929).

This followup [issue](https://github.com/airbytehq/airbyte-internal-issues/issues/547) is to allow for users to specify which email they want the notifications to go to.

## How
A ticket exists to allow users to set up a customer.io notification configuration, similar to how it works with slack notifications. Because this behavior does not exist yet, a separate method has been created to make sure that these notifications are sent even without a configuration being set. These notifications are sent to the email associated with the workspace of the connection being disabled.

## To do
- [x] send warning when there has been only failing connections for the past `maxDaysOfOnlyFailedJobs / 2` days

## Recommended reading order
1. `airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/JobNotifier.java`
2. `airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java`

### CODE REVIEW

**Comment by terencecho** [comment]
I wanted to point out for the reviewers that the warning notifications are meant to be sent out when they have reached 50% of the failure limit, so 50 out of 100 consecutive failure OR 7 days of straight failure, halfway to the 14 days where the auto disable kicks in.

For the consecutive failures, we can just warn once when the number of failures is exactly at 50, but for the days of straight failures, the current implementation does keep a state of if a warning has been sent. This means currently, we send a warning after each failure after 7 days of straight failures, which is not idea.

**Review by pmossman — COMMENTED** [review]
Took a read through this, I think the logic for the most part makes sense, though it is pretty convoluted and a bit hard to follow. I wonder if readability would be helped if the large `autoDisableFailingConnection()` method were broken into discrete private method calls?

I'm also curious what happens to the temporal workflow when a connection is auto-disabled. I see that the autoDisableFailingConnection activity updates the database status of the connection to INACTIVE, but what happens to the temporal connection manager workflow?

**Comment by terencecho** [comment]
Thanks for the comments Parker. Totally agree with the code being a bit hard to follow after I added the 7 day warning. I'll work on splitting it up and making it easier to understand. I wanted to get feedback on if I should be handling the logic in a different way.

When the connection is set to `Inactive`, the temporal connection manager workflow is still running, it's just that the next job to be scheduled is set to like 100s of years in the future (I forget the exact value), so basically stuck in a state of running, but doesn't do anything unless it receives another signal to do something.

**Comment by terencecho** [comment]
Going to move this PR back to a draft while I refactor for readability.

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java [hunk 1]
```diff
@@ -28,63 +35,104 @@
...
         final List<JobStatus> jobStatuses = jobPersistence.listJobStatusWithConnection(input.getConnectionId(), REPLICATION_TYPES,
             input.getCurrTimestamp().minus(maxDaysOfOnlyFailedJobs, ChronoUnit.DAYS));
 
+        // list of job statuses to be used in deciding if a warning should be sent; subset of `jobStatuses`
+        final List<JobStatus> jobStatusesForWarning = jobPersistence.listJobStatusWithConnection(input.getConnectionId(), REPLICATION_TYPES,
+            input.getCurrTimestamp().minus(maxDaysOfOnlyFailedJobsBeforeWarning, ChronoUnit.DAYS));
+
+        final Job lastJob;
+        final Optional<Job> optionalJob = jobPersistence.getLastReplicationJob(input.getConnectionId());
+        if (optionalJob.isPresent()) {
```
Comment (benmoriceau) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java:65]: it can be refactor in:
```
final lastJob = optionalJob.orElseThrow(() -> new Exeception())
```

**Comment by pmossman** [comment]
@terencecho @benmoriceau I wonder if we can have the connection manager workflow actually use the return value of the AutoDisableFailingConnection activity method to avoid sending the 7-day warning on each job failure. If the activity method returns output that indicates that a warning was sent, we could store that in the WorkflowState, and pass it into any subsequent calls of the AutoDisableFailingConnection method on future failures. That way the activity can know if we've already sent a warning for this connection.

Does that make sense? Let me know if I can clarify anything further about the idea

File: airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/JobNotifier.java [hunk 2]
```diff
@@ -154,6 +146,60 @@ public void notifyJobByEmail(final String reason, final String action, final Job
...
+
+      final UUID connectionId = UUID.fromString(job.getScope());
+      final StandardSourceDefinition sourceDefinition = configRepository.getSourceDefinitionFromConnection(connectionId);
+      final StandardDestinationDefinition destinationDefinition = configRepository.getDestinationDefinitionFromConnection(connectionId);
+      final String sourceConnector = String.format("%s version %s", sourceDefinition.getName(), sourceDefinition.getDockerImageTag());
+      final String destinationConnector = String.format("%s version %s", destinationDefinition.getName(), destinationDefinition.getDockerImageTag());
+      final String logUrl = connectionPageUrl + connectionId;
+      final String jobDescription = getJobDescription(job, "");
+
+      if (CONNECTION_DISABLED_NOTIFICATION.equals(action)) {
```
[the remark this answers is not in the document]
Comment (terencecho) [reply airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/JobNotifier.java:173]: agreed. I'll make a note to change these to enums once notification configurations/tables have been finalized as a part of https://github.com/airbytehq/airbyte/pull/10999

[the remark this answers is not in the document]
Comment (terencecho) [reply airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/JobNotifier.java:173]: done

File: airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/JobNotifier.java [hunk 3]
```diff
@@ -154,6 +146,60 @@ public void notifyJobByEmail(final String reason, final String action, final Job
...
+  private String getJobDescription(final Job job, final String reason) {
+    final Instant jobStartedDate = Instant.ofEpochSecond(job.getStartedAtInSecond().orElse(job.getCreatedAtInSecond()));
+    final DateTimeFormatter formatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.FULL).withZone(ZoneId.systemDefault());
+    final Instant jobUpdatedDate = Instant.ofEpochSecond(job.getUpdatedAtInSecond());
+    final Instant adjustedJobUpdatedDate = jobUpdatedDate.equals(jobStartedDate) ? Instant.now() : jobUpdatedDate;
+    final Duration duration = Duration.between(jobStartedDate, adjustedJobUpdatedDate);
+    final String durationString = formatDurationPart(duration.toDaysPart(), "day")
+        + formatDurationPart(duration.toHoursPart(), "hour")
+        + formatDurationPart(duration.toMinutesPart(), "minute")
+        + formatDurationPart(duration.toSecondsPart(), "second");
```
[the remark this answers is not in the document]
Comment (terencecho) [reply airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/JobNotifier.java:199]: done

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java (hunk 1 above)
Comment (terencecho) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java]: done

**Comment by terencecho** [comment]
makes sense to me. i definitely like that idea more than storing that info in someplace like a database.

**Comment by terencecho** [comment]
Looks like some of the tests are failing on previously known flaky tests. I think this diff is okay to be reviewed in the meanwhile.

**Comment by terencecho** [comment]
not sure if i should put the state of if a warning has been sent recently in `WorkflowState` or `WorkflowInternalState`. i just have it as a var in `ConnectionManagerWorkflowImpl` but i'm leaning on putting it in the internal state

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java [hunk 4]
```diff
@@ -265,8 +266,11 @@ private void reportFailure(final ConnectionUpdaterInput connectionUpdaterInput,
 
       if (autoDisableConnectionVersion != Workflow.DEFAULT_VERSION) {
         final AutoDisableConnectionActivityInput autoDisableConnectionActivityInput =
-            new AutoDisableConnectionActivityInput(connectionId, Instant.ofEpochMilli(Workflow.currentTimeMillis()));
-        runMandatoryActivity(autoDisableConnectionActivity::autoDisableFailingConnection, autoDisableConnectionActivityInput);
+            new AutoDisableConnectionActivityInput(connectionId, Instant.ofEpochMilli(Workflow.currentTimeMillis()),
+                workflowInternalState.getLastAutoDisableWarningTimestamp());
```
Comment (benmoriceau) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java:270]: Is it working? I don't think this properly pass from 1 wf to another (to be check).

Comment (pmossman) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java:270]: From what I can tell, `workflowInternalState` doesn't get reset between runs, we only clear particular fields of the `workflowInternalState`:

https://github.com/airbytehq/airbyte/blob/7bb4baec2546cbf3f679a1aef02db13baa6475d8/airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java#L367-L377

So I think the `lastAutoDisableWarningTimestamp` will carry forward from one run to the next. I'm not sure what happens if a temporal workflow is terminated and re-created though. We probably won't be able to do much about that. @benmoriceau what do you think?

Comment (benmoriceau) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java:270]: @terencecho @pmossman, I can show it to you. I think the main reason is that we don't pass this information in a workflow input.

Comment (terencecho) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java:270]: Had a chat with benoit. Basically when the workflow continues as new, it is re-instantiated with inputs form `connectionUpdaterInput`.

Meaning everything the `workflowInternalState` is re-instantiated and not passed in through the input. The workflow state is passed in, but we only want that passed in so we can use the event listener for testing purposes, which is why we have `workflowState.reset()`.

We could add the warning notification state into `ConnectionUpdaterInput`, but that's probably not what we want to do. I'll spend a little more time thinking of other ways to get around this. Obviously putting the state in a database is an option, but I'm also hoping to avoid that.

**Comment by terencecho** [comment]
@pmossman PTAL

**Comment by terencecho** [comment]
For extra context, I refactored the database query to also retrieve timestamps when getting status. With those timestamps, we calculate if the previous failure should have triggered a warning, if i should have, then we skip sending another warning to avoid spam.

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java [hunk 5]
```diff
@@ -28,63 +36,131 @@ public class AutoDisableConnectionActivityImpl implements AutoDisableConnectionA
...
+
+        // disable connection if only failed jobs in the past maxDaysOfOnlyFailedJobs days
+        if (numDaysSinceFirstReplicationJob >= maxDaysOfOnlyFailedJobs && successTimestamp.isEmpty()) {
+          disableConnection(input.getConnectionId(), lastJob);
+          return new AutoDisableConnectionOutput(true);
+        }
+
+        // sending warning for disable connection if only failed jobs in the past
+        // maxDaysOfOnlyFailedJobsBeforeWarning days
+        if (numDaysSinceFirstReplicationJob >= maxDaysOfOnlyFailedJobsBeforeWarning &&
```
Comment (pmossman) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java:103]: This is a bit hard to follow because of the nesting, can you pull some of the inner conditional logic out into named variables to help with readability?

Something like:
```
final boolean firstReplicationIsOldEnough = numDaysSinceFirstReplicationJob >= maxDaysOfOnlyFailedJobsBeforeWarning;
final boolean enoughDaysPassedSinceLastSuccess = successTimestamp.isEmpty()
                || getDaysSinceTimestamp(currTimestampInSeconds, successTimestamp.get()) >= maxDaysOfOnlyFailedJobsBeforeWarning);

if (firstReplicationIsOldEnough && enoughDaysPassedSinceLastSuccess) {
    ...
}
```

You should pick better variable names if you can think of them, especially if mine don't actually match the logic 😅

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java [hunk 6]
```diff
@@ -28,63 +36,131 @@ public class AutoDisableConnectionActivityImpl implements AutoDisableConnectionA
...
+
+    // get previous failed job (skipping first job since that's considered "current" job)
+    JobWithStatusAndTimestamp prevFailedJob = jobs.get(1);
+    for (int i = 2; i < jobs.size(); i++) {
+      if (prevFailedJob.getStatus() == JobStatus.FAILED)
+        break;
+      prevFailedJob = jobs.get(i);
+    }
+
+    return (successTimestamp.isEmpty()
```
Comment (pmossman) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java:148]: Similar recommendation here for using local boolean variables to aid comprehension of the nested conditional

**Review by pmossman — CHANGES_REQUESTED** [review]
@terencecho this makes sense to me! I made a recommendation to help with readability of the conditional logic in a couple places. I think once that's addressed, this should be good to go

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java (hunk 5 above)
Comment (terencecho) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java]: Thanks for the suggestions. I edited the above to help with readability.

I went with uglier, longer names like `firstReplicationOlderThanMaxDisableWarningDays` and `successOlderThanPrevFailureByMaxWarningDays` to help make things as clear as possible, but do let me know if they actually make things less clear, happy to go with your suggestions if you think so.

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java (hunk 6 above)
Comment (terencecho) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java]: same as above

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java [hunk 7]
```diff
@@ -91,18 +91,22 @@ public AutoDisableConnectionOutput autoDisableFailingConnection(final AutoDisabl
...
-        // maxDaysOfOnlyFailedJobsBeforeWarning days
-        if (numDaysSinceFirstReplicationJob >= maxDaysOfOnlyFailedJobsBeforeWarning &&
-            (successTimestamp.isEmpty()
-                || getDaysSinceTimestamp(currTimestampInSeconds, successTimestamp.get()) >= maxDaysOfOnlyFailedJobsBeforeWarning)) {
+        final boolean firstReplicationOlderThanMaxDisableWarningDays = numDaysSinceFirstReplicationJob >= maxDaysOfOnlyFailedJobsBeforeWarning;
+        final boolean successOlderThanPrevFailureByMaxWarningDays = // set to true if no previous success is found
+            noPreviousSuccess || getDaysSinceTimestamp(currTimestampInSeconds, successTimestamp.get()) >= maxDaysOfOnlyFailedJobsBeforeWarning;
+
+        // send warning if there are only failed jobs in the past maxDaysOfOnlyFailedJobsBeforeWarning days
+        // _except_ if a warning should have already been sent in the previous failure
```
Comment (pmossman) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java:108]: super nit: instead of `_except_ if a warning...`, go with `_unless_ a warning...`

**Review by pmossman — APPROVED** [review]
Nice! Definitely much easier to follow the logic now. LGTM!

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java (hunk 7 above)
Comment (terencecho) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java]: done

**Comment by terencecho** [comment]
@benmoriceau i'm going to merge this, but feel free to comment with any suggestions and I can follow up on another diff.

In case you want a tldr: didn't make any changes on the temporal side, instead decided to read timestamp info from database to determine if a warning should have already been sent to avoid spam instead of saving state of last warning sent somewhere.
